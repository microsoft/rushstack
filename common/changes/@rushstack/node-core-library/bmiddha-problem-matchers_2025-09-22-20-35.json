{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Add `FileError.getProblemMatcher()` which returns the problem matcher compatible with `IOperationExecutionResult.problemCollector` as well as VS Code, GitHub Actions",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}
