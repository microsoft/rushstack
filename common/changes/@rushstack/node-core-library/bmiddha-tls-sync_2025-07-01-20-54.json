{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Add a `Async.runWithTimeoutAsync` function that executes an async function, resolving if the specified timeout elapses first.",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}