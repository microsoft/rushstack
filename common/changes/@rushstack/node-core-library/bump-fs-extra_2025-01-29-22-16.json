{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Update fs-extra to 11.3.0.",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}