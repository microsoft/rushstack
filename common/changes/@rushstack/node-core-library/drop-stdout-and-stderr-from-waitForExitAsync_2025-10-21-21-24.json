{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Update the return type of `Executable.waitForExitAsync` to omit `stdout` and `stderr` if an `encoding` parameter isn't passed to the options object.",
      "type": "patch"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}