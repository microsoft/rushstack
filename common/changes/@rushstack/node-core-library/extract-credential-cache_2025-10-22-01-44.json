{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Add \"Objects.areDeepEqual\" and \"User.getHomeFolder\" APIs.",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}