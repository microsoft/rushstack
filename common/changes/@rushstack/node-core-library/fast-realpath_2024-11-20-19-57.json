{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Add `RealNodeModulePathResolver` class to get equivalent behavior to `realpath` with fewer system calls (and therefore higher performance) in the typical scenario where the only symlinks in the repository are inside of `node_modules` folders and are links to package folders.",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}