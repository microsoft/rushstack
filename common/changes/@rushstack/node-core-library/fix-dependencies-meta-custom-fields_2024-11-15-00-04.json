{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Extend IDependenciesMetaTable interface to support custom properties in \"dependenciesMeta\" entries",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}