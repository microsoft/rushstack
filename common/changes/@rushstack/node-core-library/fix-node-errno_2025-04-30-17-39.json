{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Fix a bug in `FileSystem.isErrnoException` that failed to identify errors if the underlying method was invoked using only a file descriptor, e.g. for `fs.readSync`.",
      "type": "patch"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}