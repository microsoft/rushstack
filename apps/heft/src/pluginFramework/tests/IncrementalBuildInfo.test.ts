// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import path from 'path';

import { Path } from '@rushstack/node-core-library';

import {
  serializeBuildInfo,
  deserializeBuildInfo,
  type IIncrementalBuildInfo,
  type ISerializedIncrementalBuildInfo
} from '../IncrementalBuildInfo';

const posixBuildInfo: IIncrementalBuildInfo = {
  configHash: 'foobar',
  inputFileVersions: new Map([
    ['/a/b/c/file1', '1'],
    ['/a/b/c/file2', '2']
  ]),
  fileDependencies: new Map([
    ['/a/b/c/output1', ['/a/b/c/file1']],
    ['/a/b/c/output2', ['/a/b/c/file1', '/a/b/c/file2']]
  ])
};

const win32BuildInfo: IIncrementalBuildInfo = {
  configHash: 'foobar',
  inputFileVersions: new Map([
    ['A:\\b\\c\\file1', '1'],
    ['A:\\b\\c\\file2', '2']
  ]),
  fileDependencies: new Map([
    ['A:\\b\\c\\output1', ['A:\\b\\c\\file1']],
    ['A:\\b\\c\\output2', ['A:\\b\\c\\file1', 'A:\\b\\c\\file2']]
  ])
};

const posixBasePath: string = '/a/b/temp';
const win32BasePath: string = 'A:\\b\\temp';

function posixToPortable(absolutePath: string): string {
  return path.posix.relative(posixBasePath, absolutePath);
}
function portableToPosix(portablePath: string): string {
  return path.posix.resolve(posixBasePath, portablePath);
}

function win32ToPortable(absolutePath: string): string {
  return Path.convertToSlashes(path.win32.relative(win32BasePath, absolutePath));
}
function portableToWin32(portablePath: string): string {
  return path.win32.resolve(win32BasePath, portablePath);
}

describe(serializeBuildInfo.name, () => {
  it('Round trips correctly (POSIX)', () => {
    const serialized: ISerializedIncrementalBuildInfo = serializeBuildInfo(posixBuildInfo, posixToPortable);

    const deserialized: IIncrementalBuildInfo = deserializeBuildInfo(serialized, portableToPosix);

    expect(deserialized).toEqual(posixBuildInfo);
  });

  it('Round trips correctly (Win32)', () => {
    const serialized: ISerializedIncrementalBuildInfo = serializeBuildInfo(win32BuildInfo, win32ToPortable);

    const deserialized: IIncrementalBuildInfo = deserializeBuildInfo(serialized, portableToWin32);

    expect(deserialized).toEqual(win32BuildInfo);
  });

  it('Converts (POSIX to Win32)', () => {
    const serialized: ISerializedIncrementalBuildInfo = serializeBuildInfo(posixBuildInfo, posixToPortable);

    const deserialized: IIncrementalBuildInfo = deserializeBuildInfo(serialized, portableToWin32);

    expect(deserialized).toEqual(win32BuildInfo);
  });

  it('Converts (Win32 to POSIX)', () => {
    const serialized: ISerializedIncrementalBuildInfo = serializeBuildInfo(win32BuildInfo, win32ToPortable);

    const deserialized: IIncrementalBuildInfo = deserializeBuildInfo(serialized, portableToPosix);

    expect(deserialized).toEqual(posixBuildInfo);
  });

  it('Has expected serialized format', () => {
    const serializedPosix: ISerializedIncrementalBuildInfo = serializeBuildInfo(
      posixBuildInfo,
      posixToPortable
    );
    const serializedWin32: ISerializedIncrementalBuildInfo = serializeBuildInfo(
      win32BuildInfo,
      win32ToPortable
    );

    expect(serializedPosix).toMatchSnapshot('posix');
    expect(serializedWin32).toMatchSnapshot('win32');

    expect(serializedPosix).toEqual(serializedWin32);
  });
});
