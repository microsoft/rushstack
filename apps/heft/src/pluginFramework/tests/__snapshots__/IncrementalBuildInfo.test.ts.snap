// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializeBuildInfo Has expected serialized format: posix 1`] = `
Object {
  "configHash": "foobar",
  "fileDependencies": Object {
    "../c/output1": Array [
      0,
    ],
    "../c/output2": Array [
      0,
      1,
    ],
  },
  "inputFileVersions": Object {
    "../c/file1": "1",
    "../c/file2": "2",
  },
}
`;

exports[`serializeBuildInfo Has expected serialized format: win32 1`] = `
Object {
  "configHash": "foobar",
  "fileDependencies": Object {
    "../c/output1": Array [
      0,
    ],
    "../c/output2": Array [
      0,
      1,
    ],
  },
  "inputFileVersions": Object {
    "../c/file1": "1",
    "../c/file2": "2",
  },
}
`;
